### ISSUE TITLE
Create a dedicated evaluation object/module

### ISSUE THREAD

**Issue by daxenberger** [dkpro/dkpro-tc#50]
Originally reported on Google Code with ID 50

```
Some reports might produce mandatory output which other tasks or reports (potentially)
depend on. This is the case for some of the reports which create the results of the
test task (f-scores etc.). We should think about moving this (core) functionality to
their an own task.
```

Reported by `daxenberger.j` on 2013-09-17 16:41:45

Edit by Tobias Horsmann - ToDo list:
- [x] make sure that each MLAdapter/Framework creates id2outcome files (TestTasks)
- [x] delete old (ML framework dependent) results reports (TestTasks)
- [x] adapt all higher(batch)-level reports to work with new evaluation mode (in particular, Issue #113)
- [x] switch all demos to new evaluation mode
- [x] implement missing measures (the wish list is probably long, but for the beginning, the most important ones would probably be related to regression, which are completely missing atm)

**Comment by Horsmann** [comment]
@daxenberger Is there anything left to do or should we move this one to 0.9.0 ?

**Comment by daxenberger** [comment]
move to the next milestone. this hasn't been tested and integrated properly yet.

**Comment by Horsmann** [comment]
If you tell me what is left to do or where to start I would continue integration. Which is the next step here?

**Comment by daxenberger** [comment]
A rough roadmap:
- [x] make sure that each MLAdapter/Framework creates id2outcome files (TestTasks)
- [x] delete old (ML framework dependent) results reports (TestTasks)
- [ ] adapt all higher(batch)-level reports to work with new evaluation mode (in particular, Issue #113)
- [x] switch all demos to new evaluation mode
- [x] implement missing measures (the wish list is probably long, but for the beginning, the most important ones would probably be related to regression, which are completely missing atm)

**Comment by Horsmann** [comment]
@daxenberger I started a new branch making SvmHmm my guinea pig. Svmhmm outputs some own confusion matrix stuff. This can be removed now, too? I am not sure if it is even correct since this also untested but it should be handled by the new evaluation module now anyway?

Except for the additional files SvmHmm creates the integration is not soo hard. I can remove `SVMHMMClassificationReport` and `SVMHMMBatchCrossValidationReport` once the interface is adapted, right?

**Comment by daxenberger** [comment]
> [...] but it should be handled by the new evaluation module now anyway?

yes, in theory. I'm not sure how well this works atm, so maybe deprecate the old reports rather than removing them completely.

> I can remove SVMHMMClassificationReport and SVMHMMBatchCrossValidationReport once the interface is adapted, right?

Some here: I'd prefer to remove them from demos etc., but instead of completely deleting them, better deprecate.

**Comment by Horsmann** [comment]
What do we do with the `Mallet` module. It is `deprecated` since `0.7.0`. All changes to the API have to implemented there too. Maybe its time to remove it entirely? Who ever needs Mallet should either use 0.7.0 or 0.8.0 ?

**Comment by reckart** [comment]
Why was it deprecated?

**Comment by Horsmann** [comment]
I think the code is a bit messy some things never seemed to have worked (sequence classification) and we have Weka and Crfsuite which covers everything. If someone wants to revive it it should be reimplemented from scratch imo.

**Comment by zesch** [comment]
Very slow. Was not really usable for real problems.
Tobias Horsmann notifications@github.com schrieb am Sa., 7. Mai 2016 um
13:33:

**Comment by Horsmann** [comment]
@daxenberger all MLA are supposed to have as last entry `threshold` in the id2outcome file? What value do I set if the MLA doesn't has such a parameter?

**Comment by daxenberger** [comment]
yeah that is necessary to have a common format for all kind of learning modes. `theshold` will be ignored if multi-labeling is not applied and should thus set be set to -1 (or 0).

**Comment by Horsmann** [comment]
@daxenberger ok thx. Do we adapt Mallet too? Maybe its time to remove it entirely?

**Comment by zesch** [comment]
How much effort is it to drag it along?
Tobias Horsmann notifications@github.com schrieb am So., 8. Mai 2016 um
12:51:

**Comment by Horsmann** [comment]
About the same as for the other modules - a have day. I am more concerned to keep maintaining a module that is dead since 2 releases. Changes have been reflected for Mallet so far if they affected interfaces and made it mandatory to change it there too (otherwise Jenkins would fail) but there are no unit tests for the module. I find its whole state a bit questionable imo it is rotting code that causes effort in every change and no one(?) has any advantage of the effort spend. So I would suggest to remove it :)

**Comment by zesch** [comment]
Ok for me
Tobias Horsmann notifications@github.com schrieb am So., 8. Mai 2016 um
13:31:

**Comment by reckart** [comment]
Might be worth noting that there is a new version of mallet since a few days:

http://search.maven.org/#artifactdetails%7Ccc.mallet%7Cmallet%7C2.0.8%7Cjar

https://github.com/mimno/Mallet

**Comment by Horsmann** [comment]
What I have done so far
- [x] add Liblinear outcome2id report
- [x] add Crfsuite outcome2id report
- [x] update (non-groovy) examples/test to use new evaluation module
- [x] delete old module-depended ClassificationReports
- [x] make `*usingTCEvaluationReport` reports main reports `BatchTrainTestReport` and `BatchCrossvalidationReport`

Ok here a list of issues I am not sure how to handle
- [x] I can't fix the Groovy Test cases - support is not available for my Eclipse version @daxenberger can you fix those?
- [x] `WekaRegressionExperimentTest` triggers an exception when calling a `getLabel()` method - no regression implemented yet?
- [x]  `Mallet` module - see previous postings - is untouched at the moment
- [x] Weka has various other *Adapter classes with additionally reports - how to handle those?

**Comment by reckart** [comment]
> I can't fix the Groovy Test cases - support is not available for my Eclipse version @daxenberger can you fix those?

For curiosity, what Eclipse version are you using?

**Comment by Horsmann** [comment]
4.4 Luna this one isn't working https://marketplace.eclipse.org/content/groovygrails-tool-suite-ggts-eclipse

**Comment by reckart** [comment]
I have been using Luna before and now I am using Mars.

For Luna: http://dist.springsource.org/release/GRECLIPSE/e4.4/
For Mars: http://dist.springsource.org/snapshot/GRECLIPSE/e4.5/

See also: https://github.com/groovy/groovy-eclipse/wiki

**Comment by Horsmann** [comment]
sry, not working I tried those too. Installation fails with an exception.

**Comment by Horsmann** [comment]
I installed another Eclipse version - I don't understand was those Groovy test cases are supposed to do which makes it hard to fix it. I could need a hand for those... @daxenberger `help wanted`

**Comment by daxenberger** [comment]
> I can't fix the Groovy Test cases - support is not available for my Eclipse version @daxenberger can you fix those?

I'm compiling the groovy demos without problems under Eclipse Mars (4.5.2) with Groovy Compiler (1.8-2.4) 2.9.2.xx

> WekaRegressionExperimentTest triggers an exception when calling a getLabel() method - no regression implemented yet?

See my reponse on the mailing list - no getLabel() method for regression.

> Mallet module - see previous postings - is untouched at the moment

I am a bit reluctant to totally remove the module since Mallet has become more active recently (see above). Maybe move the module into it's own branch?

> Weka has various other *Adapter classes with additionally reports - how to handle those?

The *Prediction adapters are not needed anymore, since we have Save/LoadModel now. 
*Statistics adapters are already ported to the new evaluation mode.
Meka and Weka needs to be distinguished due to differences in single/multi-label mode.

**Comment by daxenberger** [comment]
> I installed another Eclipse version - I don't understand was those Groovy test cases are supposed to do which makes it hard to fix it. I could need a hand for those... @daxenberger help wanted

which tests do you refer to? The tests in the groovy demo module to the same as in the java module: simply execute the demos.

**Comment by Horsmann** [comment]
the `PairTwentyNewsgroupsDemo` fails with a Stanford triggered ClassCastException on my machine? 
The regression demo should fail too due to the `label()` issue.

``````
Caused by: java.lang.ClassCastException: java.util.ArrayList cannot be cast to [Ledu.stanford.nlp.util.Index;
    at edu.stanford.nlp.ie.crf.CRFClassifier.loadClassifier(CRFClassifier.java:2164)
    at edu.stanford.nlp.ie.AbstractSequenceClassifier.loadClassifier(AbstractSequenceClassifier.java:1249)
    at edu.stanford.nlp.ie.AbstractSequenceClassifier.loadClassifier(AbstractSequenceClassifier.java:1226)
    at edu.stanford.nlp.ie.crf.CRFClassifier.getClassifier(CRFClassifier.java:2278)
    at de.tudarmstadt.ukp.dkpro.core.stanfordnlp.StanfordNamedEntityRecognizer$1.produceResource(StanfordNamedEntityRecognizer.java:170)
    at de.tudarmstadt.ukp.dkpro.core.stanfordnlp.StanfordNamedEntityRecognizer$1.produceResource(StanfordNamedEntityRecognizer.java:141)```
``````

**Comment by Horsmann** [comment]
@daxenberger 
Regarding mallet: I can fork-off a new branch from `master` with the current state. I would then remove the mallet module from my `issue50` branch which removes the module once it is merged into `master`

**Comment by daxenberger** [comment]
> the PairTwentyNewsgroupsDemo fails with a Stanford triggered ClassCastException on my machine? 

ok; I'll have a look at this

**Comment by daxenberger** [comment]
> Regarding mallet: I can fork-off a new branch from master with the current state. I would then remove the mallet module from my issue50 branch which removes the module once it is merged into master

sounds good

**Comment by Horsmann** [comment]
Oookay. I set up a Jenkins job and it seems Jenkins does not have the problems I have with the Groovy experiments. I think I am quite close to merging this branch into master. Seemingly everything is working. 
I added 2 of the most easy to implement regression measures for a few simple tests. 
The other bugs will probably only show themselves when we are actually start using the module.

**Comment by Horsmann** [comment]
I merged the changes into master - an open todo on the checklist is the measure implementations (for regression). 

@daxenberger I have the `WekaFeatureValuesReport` as left-over. The report used the former `result.txt` if this report is still needed you would have to upgrade it to using the new module. Not really sure what it is suppose to do.

Maybe we start an own issue for adding the remaining measures?

**Comment by daxenberger** [comment]
> I have the WekaFeatureValuesReport as left-over.

This report hasn't been updated since about 2 releases. Can be dropped. It was originally supposed to write a report about the average value of feature for each class. This can also be done using the output of the ExtractFeaturesTask. Not sure why the result.txt was needed.

**Comment by daxenberger** [comment]
> Maybe we start an own issue for adding the remaining measures?

Agree.
